charms/istio-pilot/src/istioctl.py

```diff
@@ -0,0 +1,214 @@
+import logging
+import subprocess
+
+import yaml
+from charmed_kubeflow_chisme.lightkube.batch import delete_many
+from lightkube import Client, codecs
+
+
+class InstallFailedError(Exception):
+    pass
+
+
+class ManifestFailedError(Exception):
+    pass
+
+
+class PrecheckFailedError(Exception):
+    pass
+
+
+class UpgradeFailedError(Exception):
+    pass
+
+
+class VersionCheckError(Exception):
+    pass
```

Comment:
if you will try except them and do different treatments, then it might be the case. However, so far I do not see any try/except block for these errors :)

Reply:
but then we use only one of them per function, which means, that there is no branch required. 
So, if you raise for example `UpgradeFailedError` in all of them, nothing will be changed.

Reply:
ahh I see what you're saying.  You're saying that `Istioctl.precheck` only raises one Exception, `istioctl.upgrade` only raises one Exception, etc, so they don't need to be different types.  Yeah that's true

We can collapse them into one, although I think a single custom exception still makes catching it in the charm easier.  Would that work for you?  Maybe just `IstioctlError`?  I'll patch that now